core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala

```diff
@@ -535,7 +536,7 @@ class GroupMetadataManager(val brokerId: Int,
 
     val numExpiredOffsetsRemoved = inWriteLock(offsetExpireLock) {
       val expiredOffsets = offsetsCache.filter { case (groupTopicPartition, offsetAndMetadata) =>
-        offsetAndMetadata.expireTimestamp < startMs
+        offsetAndMetadata.expireTimestamp < startMs || !metadataCache.contains(groupTopicPartition.topicPartition.topic)
```

Comment:
The new code is just a one-liner method in MetadataCache. I am not sure if contention is a real problem, but I'd take the cost of that one line to keep the problem off the table. As a bonus, it gets around the need to grab the partitionMetadata lock while holding the offsetsExpire lock. We've had tricky deadlocks in this code before, so I'm for any opportunity to avoid holding multiple locks at the same time.